wal-scripts/bind-css/src/css_binder.rs

```diff
@@ -0,0 +1,189 @@
+use std::{
+    collections::HashSet,
+    error::Error,
+    fs::{self, OpenOptions},
+    io::Write,
+    path::{Path, PathBuf},
+};
+
+use regex::Regex;
+
+pub struct CssBinder {
+    output_path: PathBuf,
+    bound_stylesheets: HashSet<String>,
+}
+
+impl CssBinder {
+    pub fn new(output_path: &str) -> CssBinder {
+        let mut binder = CssBinder {
+            output_path: PathBuf::from(output_path),
+            bound_stylesheets: HashSet::new(),
+        };
+        binder.reset();
+        binder
+    }
+
+    // Public methods
+    pub fn reset(&mut self) {
+        if self.output_path.exists() {
+            fs::remove_file(&self.output_path).unwrap();
+        }
+        self.bound_stylesheets.clear();
+    }
+    pub fn bind_stylesheet(&mut self, path: PathBuf) -> Result<(), Box<dyn Error>> {
+        if !path.exists() {
+            panic!("found path does not exist: {}", path.display());
+        }
+
+        let stylesheet_name = Self::get_component_name_from_path(&path);
+
+        if !self.bound_stylesheets.insert(stylesheet_name.to_owned()) {
+            panic!("error inserting stylesheet name: {}", stylesheet_name);
+        }
+
+        let mut stylesheet_str = Self::read_file(path)?;
+        stylesheet_str = Self::skip_comments(stylesheet_str);
+        stylesheet_str = Self::collapse_whitespaces(stylesheet_str);
+
+        while !stylesheet_str.is_empty() {
+            let parsed_instruction = Self::parse_instruction(&mut stylesheet_str, &stylesheet_name);
+            self.write_to_output(parsed_instruction)?;
+        }
+
+        Ok(())
+    }
+    pub fn bind_global_stylesheet(&mut self, path: PathBuf) -> Result<(), Box<dyn Error>> {
+        if !path.exists() {
+            panic!("found path does not exist: {}", path.display());
+        }
+
+        let mut global_stylesheet_str = Self::read_file(path)?;
+        global_stylesheet_str = Self::skip_comments(global_stylesheet_str);
+        global_stylesheet_str = Self::collapse_whitespaces(global_stylesheet_str);
+        self.write_to_output(global_stylesheet_str)?;
+
+        Ok(())
+    }
+
+    // utility private methods
+    fn get_component_name_from_path(path: &Path) -> String {
+        let path_str = path.display().to_string();
+        let rgx = Regex::new(r"/|\\").unwrap();
+        let parts = rgx.split(&path_str);
+        parts
+            .last()
+            .unwrap()
+            .strip_suffix(".wal.css")
+            .unwrap()
+            .to_owned()
+    }
+    fn trim_front_whitespaces(str: &mut String) {
+        while str.starts_with(|c: char| c.is_whitespace()) {
```

Comment:
consider using trim start or if you want to mutate existing String "if let Some(first_non_whitespace) = s.find(|c: char| !c.is_whitespace()) {
        *s = s[first_non_whitespace..].to_string();
    } else {
        s.clear();
    }"